### PR TITLE
Glossary update, Closes Issue #16891

### PR DESCRIPTION
**Pull request by refactormyself**
Glossary Terms "Filter" and "Query" were added with details as described here https://www.elastic.co/guide/en/elasticsearch/guide/current/_queries_and_filters.html

### CODE REVIEW

**Review by refactormyself — COMMENTED** [review]
License now signed

File: docs/reference/glossary.asciidoc [hunk 1]
```diff
@@ -105,6 +115,19 @@
   +
   See also <<glossary-routing,routing>>
 
+[[glossary-query]] query ::
+
+  Query refers to all queries which not only determine if a document matches, 
+  but also calculate how well the document matches. This calculation is refered 
+  to as scoring, hence these queries are also known as "scoring queries". 
+  A scoring query calculates how relevant each document is to the query, and assigns 
+  it a relevance _score, which is later used to sort matching documents by relevance.
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: nit: use backticks for `_score`

File: docs/reference/glossary.asciidoc [hunk 2]
```diff
@@ -61,6 +61,16 @@
   `object`. The mapping also allows you to define (amongst other things)
   how the value for a field should be analyzed.
 
+[[glossary-filter]] filter ::
+
+  A filter is kind of query known as a "non-scoring" query. It does not give a score,
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: This is correct, but could you make it clearer from the beginning that a "filter" is a normal query, just one that doesn't score? You already say this, but in the current order the definition isn't as clear as it could be. I think this just needs a little rephrasing.

Comment (cbuescher) [reply docs/reference/glossary.asciidoc]: also: "is a kind" if you leave that part in anywhere.

File: docs/reference/glossary.asciidoc [hunk 3]
```diff
@@ -61,6 +61,16 @@
...
+  it is only concerned about answering the question - "Does this document match?". 
+  The answer is always a simple, binary yes|no. This kind of query is said to be made
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: In the context of the glossary I would spell "yes|no" out and say `yes` or `no`.

File: docs/reference/glossary.asciidoc [hunk 4]
```diff
@@ -61,6 +61,16 @@
   `object`. The mapping also allows you to define (amongst other things)
   how the value for a field should be analyzed.
 
+[[glossary-filter]] filter ::
+
+  A filter is kind of query known as a "non-scoring" query. It does not give a score,
+  it is only concerned about answering the question - "Does this document match?". 
+  The answer is always a simple, binary yes|no. This kind of query is said to be made 
+  in a "filtering" context, hence it is called a filter. Filtering queries/Filters are
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: Maybe explain "filtering" context, e.g. the "filter" section in a boolean query.

Comment (cbuescher) [reply docs/reference/glossary.asciidoc]: Why the "/" here? I'd simply talk about "filters" here.

File: docs/reference/glossary.asciidoc [hunk 5]
```diff
@@ -61,6 +61,16 @@
   `object`. The mapping also allows you to define (amongst other things)
   how the value for a field should be analyzed.
 
+[[glossary-filter]] filter ::
+
+  A filter is kind of query known as a "non-scoring" query. It does not give a score,
+  it is only concerned about answering the question - "Does this document match?". 
+  The answer is always a simple, binary yes|no. This kind of query is said to be made 
+  in a "filtering" context, hence it is called a filter. Filtering queries/Filters are 
+  simple checks for set inclusion/exclusion. The goal of filtering is to reduce the
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: "inclusion or exclusion"

File: docs/reference/glossary.asciidoc [hunk 6]
```diff
@@ -61,6 +61,16 @@
...
 
+[[glossary-filter]] filter ::
+
+  A filter is kind of query known as a "non-scoring" query. It does not give a score,
+  it is only concerned about answering the question - "Does this document match?". 
+  The answer is always a simple, binary yes|no. This kind of query is said to be made 
+  in a "filtering" context, hence it is called a filter. Filtering queries/Filters are 
+  simple checks for set inclusion/exclusion. The goal of filtering is to reduce the 
+  number of documents that have to be examined, as it is in the case of 
+  <<glossary-query,scoring queries>>.
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: I don't undestand the reference to scoring queries in this sentence, maybe you can explain or rephrase to make this clearer?

File: docs/reference/glossary.asciidoc [hunk 7]
```diff
@@ -105,6 +115,19 @@
   +
   See also <<glossary-routing,routing>>
 
+[[glossary-query]] query ::
+
+  Query refers to all queries which not only determine if a document matches,
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: You start by contrasting queries with something else (filters I guess), but maybe its possible to start the definition in general? Maybe talk about queries as that what defines a search request and then later differentiate between queries that score and queries that don't score (filters)? Not sure about the best phrasing for this, I think the current formulations are already quiete good, maybe also just change the ordering a bit?

File: docs/reference/glossary.asciidoc [hunk 8]
```diff
@@ -105,6 +115,19 @@
...
+[[glossary-query]] query ::
+
+  Query refers to all queries which not only determine if a document matches, 
+  but also calculate how well the document matches. This calculation is refered 
+  to as scoring, hence these queries are also known as "scoring queries". 
+  A scoring query calculates how relevant each document is to the query, and assigns 
+  it a relevance _score, which is later used to sort matching documents by relevance. 
+  This concept of relevance is well suited to full-text search, where there is seldom a 
+  completely “correct” answer. These queries are more heavier than 
+  <<glossary-filter,filters/non scoring queries>> and their query results are not cacheable.
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: I would either use "filters" or "non-scoring queries" here, probably the later.

File: docs/reference/glossary.asciidoc [hunk 9]
```diff
@@ -105,6 +115,19 @@
...
 
+[[glossary-query]] query ::
+
+  Query refers to all queries which not only determine if a document matches, 
+  but also calculate how well the document matches. This calculation is refered 
+  to as scoring, hence these queries are also known as "scoring queries". 
+  A scoring query calculates how relevant each document is to the query, and assigns 
+  it a relevance _score, which is later used to sort matching documents by relevance. 
+  This concept of relevance is well suited to full-text search, where there is seldom a 
+  completely “correct” answer. These queries are more heavier than
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: Not sure if "more heavier" is correct (I'm not a native speaker) but it sounds like using "heavier" would be enough. Maybe "heavier" could also be exchanged for something that makes it clearer that we are talking about performance here.

**Review by cbuescher — CHANGES_REQUESTED** [review]
Hi @refactormyself,

thanks for picking up the issue and submitting this PR, greatly appreciated. I left a couple of small comments where I think the phrasing could be improved a bit but please take these only as suggestions. Since this is going to end up in the glossary I think its worth to try to find a good, concise definition of the terms, so hope you don't mind some of the comments. Also appreciate anybody elses opinions on this.

File: docs/reference/glossary.asciidoc (hunk 4 above)
Comment (refactormyself) [reply docs/reference/glossary.asciidoc]: "Filtering queries/Filters" is used to emphasis that Filtering queries = Filters

File: docs/reference/glossary.asciidoc (hunk 6 above)
Comment (refactormyself) [reply docs/reference/glossary.asciidoc]: The goal of filtering is to reduce the number of documents that have to be examined, **instead of what happens** in the case of  <<glossary-query,scoring queries>>.

The reference is also meant to provide a link to the reader for comparism

**Comment by refactormyself** [comment]
@cbuescher I hope I have addressed your comments

File: docs/reference/glossary.asciidoc [hunk 10]
```diff
@@ -61,6 +61,16 @@
   `object`. The mapping also allows you to define (amongst other things)
   how the value for a field should be analyzed.
 
+[[glossary-filter]] filter ::
+
+  A filter is a query. It is a kind of query which does not give a score, so it is
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: Maybe you can join the first two sentences to make it more fluent, e.g. "A filter is a <<glossary-query,scoring queries>> that ..." and then join the second sentence. 
Maybe use something like "produces a score" or "scores documents" instead of "give a score"?

File: docs/reference/glossary.asciidoc [hunk 11]
```diff
@@ -61,6 +61,16 @@
   `object`. The mapping also allows you to define (amongst other things)
   how the value for a field should be analyzed.
 
+[[glossary-filter]] filter ::
+
+  A filter is a query. It is a kind of query which does not give a score, so it is 
+  known as a "non-scoring" query.It is only concerned about answering the question -
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: nit: whitespace between "query.It". Although the difference won't show up in the docs I think, it improves readability here.

File: docs/reference/glossary.asciidoc [hunk 12]
```diff
@@ -61,6 +61,16 @@
   `object`. The mapping also allows you to define (amongst other things)
   how the value for a field should be analyzed.
 
+[[glossary-filter]] filter ::
+
+  A filter is a query. It is a kind of query which does not give a score, so it is 
+  known as a "non-scoring" query.It is only concerned about answering the question - 
+  "Does this document match?". The answer is always a simple, binary yes or no. This kind of query is said to be made
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: Could you wrap this (and the following lines) roughly at the same position as the previous lines? We have different line lengths in the docs throughout the documentation but at least in one paragraph it would be nice if it was roughly similar. Maybe it makes sense to do the wrapping in the end just before the PR is ready to be merged though, to just go through the trouble once.

File: docs/reference/glossary.asciidoc [hunk 13]
```diff
@@ -61,6 +61,16 @@
   `object`. The mapping also allows you to define (amongst other things)
   how the value for a field should be analyzed.
 
+[[glossary-filter]] filter ::
+
+  A filter is a query. It is a kind of query which does not give a score, so it is 
+  known as a "non-scoring" query.It is only concerned about answering the question - 
+  "Does this document match?". The answer is always a simple, binary yes or no. This kind of query is said to be made 
+  in a "filtering" context (e.g. Is the created date in the range 2013 - 2014?), hence it is called a filter. Filters are
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: Maybe we should link to https://www.elastic.co/guide/en/elasticsearch/reference/current/query-filter-context.html when mentioning the filter context. Also it would be great if it would be clearer that "filter context" defines specific locations of a query clause in a combination of queries. If this is too much detail for the glossary, I think linking to the "filter context" paragraph and removing the details here is fine as well.

File: docs/reference/glossary.asciidoc [hunk 14]
```diff
@@ -61,6 +61,16 @@
   `object`. The mapping also allows you to define (amongst other things)
   how the value for a field should be analyzed.
 
+[[glossary-filter]] filter ::
+
+  A filter is a query. It is a kind of query which does not give a score, so it is 
+  known as a "non-scoring" query.It is only concerned about answering the question - 
+  "Does this document match?". The answer is always a simple, binary yes or no. This kind of query is said to be made 
+  in a "filtering" context (e.g. Is the created date in the range 2013 - 2014?), hence it is called a filter. Filters are 
+  simple checks for set inclusion or exclusion. The goal of filtering is to reduce the
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: > "The goal of filtering is to reduce the number of documents"

This might be true most of the time, but e.g. `constant_score` isn't really about reducing the document set, its just about saying you don't care about the score. I'd be okay with adding "most of the time" somewhere, maybe also leave out the "instead of..." part, since reducing document set size might also happen in scoring queries.

File: docs/reference/glossary.asciidoc [hunk 15]
```diff
@@ -105,6 +115,20 @@
   +
   See also <<glossary-routing,routing>>
 
+[[glossary-query]] query ::
+
+  A query is the basic component of a search. A search can be defined by one or more queries 
+  which can be mixed and matched in endless combinations. The term Query refers to all queries
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: I'd scratch this statement, I think of "Query" as a superset of both "scoring queries" and filters, so I wouldn't introduce a distinction here.

File: docs/reference/glossary.asciidoc [hunk 16]
```diff
@@ -105,6 +115,20 @@
   +
   See also <<glossary-routing,routing>>
 
+[[glossary-query]] query ::
+
+  A query is the basic component of a search. A search can be defined by one or more queries 
+  which can be mixed and matched in endless combinations. The term Query refers to all queries 
+  which not only determine if a document matches, but also calculate how well the document matches.
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: Maybe say "Queries that calculate how well the document matches are also known as "scoring queries", as oposed to queries that only determine if a document matches, which are also called <<glossary-filter,filters>>.

File: docs/reference/glossary.asciidoc [hunk 17]
```diff
@@ -105,6 +115,20 @@
   +
   See also <<glossary-routing,routing>>
 
+[[glossary-query]] query ::
+
+  A query is the basic component of a search. A search can be defined by one or more queries 
+  which can be mixed and matched in endless combinations. The term Query refers to all queries 
+  which not only determine if a document matches, but also calculate how well the document matches. 
+  This calculation is refered to as scoring, hence these queries are also known as "scoring queries". 
+  A scoring query calculates how relevant each document is to the query, and assigns
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: I would leave out the notion of "relevance" here, this is opening yet another term that we'd need to refine more carefully I think. I would suggest using "how well a document mathces a query", "score" and "sort by score" instead.

File: docs/reference/glossary.asciidoc [hunk 18]
```diff
@@ -105,6 +115,20 @@
...
 
+[[glossary-query]] query ::
+
+  A query is the basic component of a search. A search can be defined by one or more queries 
+  which can be mixed and matched in endless combinations. The term Query refers to all queries 
+  which not only determine if a document matches, but also calculate how well the document matches. 
+  This calculation is refered to as scoring, hence these queries are also known as "scoring queries". 
+  A scoring query calculates how relevant each document is to the query, and assigns 
+  it a relevance score, which is later used to sort matching documents by relevance. 
+  This concept of relevance is well suited to full-text search, where there is seldom a
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: Just leave out "concept of relevance" if you agree with my former proposal.

File: docs/reference/glossary.asciidoc [hunk 19]
```diff
@@ -105,6 +115,20 @@
...
+[[glossary-query]] query ::
+
+  A query is the basic component of a search. A search can be defined by one or more queries 
+  which can be mixed and matched in endless combinations. The term Query refers to all queries 
+  which not only determine if a document matches, but also calculate how well the document matches. 
+  This calculation is refered to as scoring, hence these queries are also known as "scoring queries". 
+  A scoring query calculates how relevant each document is to the query, and assigns 
+  it a relevance score, which is later used to sort matching documents by relevance. 
+  This concept of relevance is well suited to full-text search, where there is seldom a 
+  completely “correct” answer. These queries are takes more resources than
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: Maybe use "Scoring queries" at the beginning of this sentence here.

File: docs/reference/glossary.asciidoc [hunk 20]
```diff
@@ -105,6 +115,20 @@
...
+  which can be mixed and matched in endless combinations. The term Query refers to all queries 
+  which not only determine if a document matches, but also calculate how well the document matches. 
+  This calculation is refered to as scoring, hence these queries are also known as "scoring queries". 
+  A scoring query calculates how relevant each document is to the query, and assigns 
+  it a relevance score, which is later used to sort matching documents by relevance. 
+  This concept of relevance is well suited to full-text search, where there is seldom a 
+  completely “correct” answer. These queries are takes more resources than 
+  <<glossary-filter,non scoring queries>> and their query results are not cacheable. 
+  As a general rule, use query clauses for full-text search or for any condition that should 
+  affect the relevance score, and use filters for everything else.
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: maybe user "condition that requires scoring" if you agree on leaving you the notion of relevance here.

**Review by cbuescher — CHANGES_REQUESTED** [review]
@refactormyself thanks for the changes, I did another round of review and left yet another set of comments.

The PR looks great in general, sorry if the comments I left feel a bit picky, but I'd like to be a bit more careful about definitions in the glossary. Maybe I'll also ping another member of the team to take another look if they agree with me and to get another pair of opinions.

File: docs/reference/glossary.asciidoc (hunk 11 above)
Comment (refactormyself) [reply docs/reference/glossary.asciidoc]: Thanks, @cbuescher I will work on your review. As a beginner to Elasticsearch myself, I appreciate the importance of getting the glossary right. I'll be happy to get it right.

File: docs/reference/glossary.asciidoc (hunk 15 above)
Comment (refactormyself) [reply docs/reference/glossary.asciidoc]: Please, I will like to clarify this. So you mean whenever the term **Query** is used it can either mean "filters" or "scoring queries". So, unless it is clarified it can be either (or I guess, even both).

Comment (cbuescher) [reply docs/reference/glossary.asciidoc]: Others might disagree, but my understanding is that we only have "queries" in elasticsearch now (there used to be special filters in earlier versions). Some queries, if used in a "filtering context" as described in the link mentioned above, are also refered to as "filters" or "non-scoring queries" as oposed to "scoring queries" (when the queries are producing scores for documents). Whether something is a filter or not depends on their location. e.g. a `term_query` is generally speaking alway a query, but when it is used e.g. in a `boolean` queries `filter` clause it is also refered to as a filter in our documentation.

Comment (cbuescher) [reply docs/reference/glossary.asciidoc]: The above explanation might be a bit verbose for the glossary however, so making it a bit more succinct would be the  goal here.

**Comment by refactormyself** [comment]
@cbuescher I made the changes, please review

File: docs/reference/glossary.asciidoc [hunk 21]
```diff
@@ -54,19 +54,26 @@
   pairs. The value can be a simple (scalar) value (eg a string, integer,
   date), or a nested structure like an array or an object. A field is
   similar to a column in a table in a relational database.
-  +
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: I'd prefer to keep these lines in (also please revert the following deletes). They prevent paragraph breaks but make the source file more readable.

File: docs/reference/glossary.asciidoc [hunk 22]
```diff
@@ -54,19 +54,26 @@
...
   of data that can be stored in that field, eg `integer`, `string`,
   `object`. The mapping also allows you to define (amongst other things)
   how the value for a field should be analyzed.
 
+[[glossary-filter]] filter ::
+
+  A filter is a non-scoring <<glossary-query,query>>, meaning that it does not score documents.
+  It is only concerned about answering the question - "Does this document match?". 
+  The answer is always a simple, binary yes or no. This kind of query is said to be made 
+  in a https://www.elastic.co/guide/en/elasticsearch/reference/current/query-filter-context.html["filtering" context],
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: I think you can use a relative link like <<query-filter-context,filter context>> here.

File: docs/reference/glossary.asciidoc [hunk 23]
```diff
@@ -95,27 +100,33 @@
...
   You cannot change the number of primary shards in an index, once the
   index is created.
-  +
   See also <<glossary-routing,routing>>
 
+[[glossary-query]] query ::
+
+  A query is the basic component of a search. A search can be defined by one or more queries 
+  which can be mixed and matched in endless combinations. While <<glossary-filter,filters>> are
+  queries that only determine if a document matches, those queries that also calculate or score
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: "calculate or score" sound like two separate things, I'd just use one, probably prefer "... also calculate how wel..."

File: docs/reference/glossary.asciidoc [hunk 24]
```diff
@@ -95,27 +100,33 @@
...
   index is created.
-  +
   See also <<glossary-routing,routing>>
 
+[[glossary-query]] query ::
+
+  A query is the basic component of a search. A search can be defined by one or more queries 
+  which can be mixed and matched in endless combinations. While <<glossary-filter,filters>> are
+  queries that only determine if a document matches, those queries that also calculate or score 
+  how well the document matches are known as "scoring queries". A scoring query calculates how
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: I'd scratch "calculates how well a document matches a query" here becauce this is already mentioned in the previous sentence. Maybe just use "Those queries assigns a score, which ..." and also leave out "... by score" at the end, too much repetition.

File: docs/reference/glossary.asciidoc [hunk 25]
```diff
@@ -95,27 +100,33 @@
...
-  +
   See also <<glossary-routing,routing>>
 
+[[glossary-query]] query ::
+
+  A query is the basic component of a search. A search can be defined by one or more queries 
+  which can be mixed and matched in endless combinations. While <<glossary-filter,filters>> are
+  queries that only determine if a document matches, those queries that also calculate or score 
+  how well the document matches are known as "scoring queries". A scoring query calculates how 
+  well a document mathces a query, and assigns it a score, which is later used to sort matching
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: "matches", but not necessary if you scratch the whole part

**Review by cbuescher — CHANGES_REQUESTED** [review]
@refactormyself thanks a lot, this looks quiet good to me already. I left a couple of small remarks still for various style reasons and to avoid repetitions, but the two sections are getting close to being ready I think.

File: docs/reference/glossary.asciidoc (hunk 22 above)
Comment (refactormyself) [reply docs/reference/glossary.asciidoc]: I tried it is not working

File: docs/reference/glossary.asciidoc (hunk 21 above)
Comment (refactormyself) [reply docs/reference/glossary.asciidoc]: Sorry, please I don't understand your point her

**Comment by refactormyself** [comment]
@cbuescher hope I placed the "+" right back

File: docs/reference/glossary.asciidoc (hunk 21 above)
Comment (cbuescher) [reply docs/reference/glossary.asciidoc]: The point was that the added free line in the documentation source is more readable as it separates the following paragraph visually, while the "+" prevents this from happening in the final page. I think this was intentional and shouldn't be changed as part of this PR, so please revert these changes.

File: docs/reference/glossary.asciidoc [hunk 26]
```diff
@@ -128,7 +146,6 @@
   the ID of the document or, if the document has a specified parent
   document, from the ID of the parent document (to ensure that child and
   parent documents are stored on the same shard).
-  +
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: please also revert this change

File: docs/reference/glossary.asciidoc [hunk 27]
```diff
@@ -143,7 +160,6 @@
   Other than defining the number of primary and replica shards that an
   index should have, you never need to refer to shards directly.
   Instead, your code should deal only with an index.
-  +
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: please also revert this change

File: docs/reference/glossary.asciidoc (hunk 22 above)
Comment (cbuescher) [reply docs/reference/glossary.asciidoc]: I also just tried, works for me. What are you doing to check the resulting page? Have you tried building the whole reference doc, e.g. by using something like `build_docs.pl --doc docs/reference/index.asciidoc --out ~/temp/asciidoc/  --open` where build_docs.pl points to my local checkout of https://github.com/elastic/docs?

**Review by cbuescher — CHANGES_REQUESTED** [review]
@refactormyself thanks for the last round of updates. I like the text much better now, I think the only think left is trying to revert all the remaining formatting changes not related to the two new sections and maybe try using the relative link again. Let me know if you have issues with that, I tried it and it works for me as I mentioned in the comment.

File: docs/reference/glossary.asciidoc [hunk 28]
```diff
@@ -61,6 +61,15 @@
   `object`. The mapping also allows you to define (amongst other things)
   how the value for a field should be analyzed.
 
+[[glossary-filter]] filter ::
+
+  A filter is a non-scoring <<glossary-query,query>>, meaning that it does not score documents.
+  It is only concerned about answering the question - "Does this document match?". 
+  The answer is always a simple, binary yes or no. This kind of query is said to be made 
+  in a <<https://www.elastic.co/guide/en/elasticsearch/reference/current/query-filter-context.html,"filtering" context>>,
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: Unfortunately this way of adding the link doesn't work, when I build the page I get a link validity error. You should use a link Id like in the other cases in this paragraph. In this case `<<query-filter-context,filter context>>` should work. I would also leave out the quotes and change the link text to "filter context" since that is the what it is called in the page that is refered to.

**Review by cbuescher — CHANGES_REQUESTED** [review]
Almost there ;-). The link doesn't work yet, unfortunately, I left a comment how to fix it. After that I think this is ready to get merged. Thanks for sticking with this until now.

File: docs/reference/glossary.asciidoc (hunk 28 above)
Comment (refactormyself) [reply docs/reference/glossary.asciidoc]: Check it and  see now, but it is not looking good on my end

Comment (cbuescher) [reply docs/reference/glossary.asciidoc]: What do you mean by "not looking good on my end"? The current update still contains the absolute link and that will break our docs build, so this will need changing unfortunately. What do you run currently to check if the docs are built correctly?

Comment (refactormyself) [reply docs/reference/glossary.asciidoc]: I think it is okay now, I didn't pay attention to what I doing wrong until you mentioned about the "absolute link" again. Hope is fixed now.

File: docs/reference/glossary.asciidoc [hunk 29]
```diff
@@ -66,7 +66,7 @@
   A filter is a non-scoring <<glossary-query,query>>, meaning that it does not score documents.
   It is only concerned about answering the question - "Does this document match?". 
   The answer is always a simple, binary yes or no. This kind of query is said to be made 
-  in a <<https://www.elastic.co/guide/en/elasticsearch/reference/current/query-filter-context.html,filter context>>, 
+  in a <<query-dsl/query_filter_context#,filter context>>,
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: This still gives an error when building the full reference docs (e.g. using `build_docs.pl --doc docs/reference/index.asciidoc --out ~/temp/asciidoc/  --open`):

```
/Users/xxx/temp/asciidoc/index.xml:78053: element link: validity error : Syntax of value for attribute linkend of link is not valid
  in a <link linkend="query-dsl/query_filter_context#">filter context</link>,
                                                      ^
/Users/xxx/temp/asciidoc/index.xml:78053: element link: validity error : IDREF attribute linkend references an unknown ID "query-dsl/query_filter_context#"
```
Which is why I was asking you earlier how you build the docs and what errors you get with the version I suggested (<<query-filter-context,filter context>>). Can you either tell me that so I can help or simply use that link instead? .

**Comment by cbuescher** [comment]
@refactormyself sorry, I'm still having trouble with that link. Have you built the docs yourself? If so, how? What was the problem with the version I suggested?

**Comment by cbuescher** [comment]
@refactormyself any news here? I think this is very close, needs only the link to work correctly.

**Comment by cbuescher** [comment]
> I just can't correlate the link between https://github.com/elastic/elasticsearch/tree/master/docs and this docs repo here https://github.com/elastic/docs.
> 
> My guess is that the latter is generated from the former, but how? I am definitely missing out something. Please help.

Sure, happy to help. 
The first url you mentioned (https://github.com/elastic/elasticsearch/tree/master/docs) points to the elasticsearch documentation source files that are kept together with the code in the main elasticsearch open source repository. This is where you already made changes to the `glossary.asciidoc` document. 
The second location (https://github.com/elastic/docs) points to a repository that contains both tools for building the documentation as well as some of the artifacts produced by that build process. That why there's so much html stuff in there.
To check if changes you made work, please refer to the [README.asciidoc](https://github.com/elastic/docs/blob/master/README.asciidoc) in the later repo. For example, I have this repo checked out locally and have a shortcut link pointing to the `build_docs.pl` script that is located in the root folder of that repo. When I e.g. try to build the glossary page alone locally I do `build_docs.pl --doc docs/reference/glossary.asciidoc --out ~/temp/asciidoc/ --open --lenient`. `--open` opens the document in the browser once build and `--lenient` supresses the link errors that I get otherwise because I'm only building a single page. With this you won't be able to check the link works correctly (because of the --lenient) but it helps in earlier phases of making small changes.
In the case of your PR, to check the link works you probably need to build the whole reference book by using something like `build_docs.pl --doc docs/reference/index.asciidoc --out ~/temp/asciidoc/ --open`. This might take a few minutes (e.g. approx. 3 minutes on my laptop) but will resolve links etc... This is where I get the above mentioned error when building from your PR branch. If you use `<<query-filter-context,filter context>>` instead of the current link, the link works when building the reference docs locally.

If you have trouble verifying using any of the above steps please let me know, otherwise I think this PR only needs correction of that "filter context" link before merging.

File: docs/reference/glossary.asciidoc [hunk 30]
```diff
@@ -61,6 +61,15 @@
   `object`. The mapping also allows you to define (amongst other things)
   how the value for a field should be analyzed.
 
+[[glossary-filter]] filter ::
+
+  A filter is a non-scoring <<glossary-query,query>>, meaning that it does not score documents.
+  It is only concerned about answering the question - "Does this document match?". 
+  The answer is always a simple, binary yes or no. This kind of query is said to be made 
+  in a https://www.elastic.co/guide/en/elasticsearch/guide/current/_queries_and_filters.html[filter context],
```
Comment (cbuescher) [inline docs/reference/glossary.asciidoc]: I think I see your problem now.
You are trying to link to a page in the elasticsearch guide, which only seems to work with an [external link syntax](https://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#links). What I said earlier was to please use an "Internal cross references" as described in the asciidoc syntax above to point to [this document](https://github.com/elastic/elasticsearch/blob/master/docs/reference/query-dsl/query_filter_context.asciidoc) in the reference itself, which works because of its inline anchor in the first line. You should be able to use `<<query-filter-context,filter context>>` for this.

**Review by cbuescher — CHANGES_REQUESTED** [review]
I think I see why the way document you are trying to link only works with an external link. I'd prefer to link a different page in reference documentation itself.

File: docs/reference/glossary.asciidoc (hunk 30 above)
Comment (refactormyself) [reply docs/reference/glossary.asciidoc]: I have tried even this `<<query_filter_context.asciidoc,filter context>>` before, now again, it is not working. It only works if both the linked and linking documents are in the same folder. Here we have `docs/reference/glossary.asciidoc` linking to `docs/reference/query-dsl/query_filter_context.asciidoc`.  I found this [issue](https://github.com/asciidoctor/asciidoctor/issues/844) on asciidoctor repo but I could not make anything of it. I tried the follow with no success :- 
`<</query-dsl/query_filter_context.asciidoc,filter context>>`  
`<<./query-dsl/query_filter_context.asciidoc,filter context>>` 
 `<<../query-dsl/query_filter_context.asciidoc,filter context>>`

I build this before pushing and it did not break any test. Must all references follow the `<<,>>` syntax ? Sorry, please I don't seem to get the problem, in this case.

Comment (refactormyself) [reply docs/reference/glossary.asciidoc]: This `<<query-filter-context,filter context>>` gives me this:
```
/temp/asciidoc/index.xml:78053: element link: validity error : IDREF attribute linkend references an unknown ID "query_filter_context"
```

I am on Windows 7, using CGwin - bash, perl, python, xsltproc installed

elasticdocs

Comment (cbuescher) [reply docs/reference/glossary.asciidoc]: Can you please push the version where you get that error?

**Review by cbuescher — CHANGES_REQUESTED** [review]

File: docs/reference/glossary.asciidoc [hunk 31]
```diff
@@ -66,7 +66,7 @@
   A filter is a non-scoring <<glossary-query,query>>, meaning that it does not score documents.
   It is only concerned about answering the question - "Does this document match?". 
   The answer is always a simple, binary yes or no. This kind of query is said to be made 
-  in a https://www.elastic.co/guide/en/elasticsearch/guide/current/_queries_and_filters.html[filter context], 
+  in a <<query_filter_context,filter context>>,
```
[the remark this answers is not in the document]
Comment (refactormyself) [reply docs/reference/glossary.asciidoc]: Okay I did it now. I guess it's working now. Sorry, I should have paid more attention to the docs. Thanks

[the remark this answers is not in the document]
Comment (cbuescher) [reply docs/reference/glossary.asciidoc]: I'm glad this works on your side now as well.

**Review by cbuescher — APPROVED** [review]
@refactormyself thanks for the latest change, this looks ready now. I'll merge this to the currently active branches now. I think it is a great addition to the glossary.
